examples/ring/vOLE/main.go

```diff
@@ -0,0 +1,406 @@
+package main
+
+import (
+	"fmt"
+	"github.com/ldsec/lattigo/v2/ring"
+	"github.com/ldsec/lattigo/v2/utils"
+	"math/big"
+	"time"
+)
+
+type benchParams struct {
+	logN   int
+	logQ   [2]int
+	n      int
+	plevel int
+	mlevel int
+}
+
+var params = []benchParams{
+	{logN: 13, logQ: [2]int{4, 60}, n: 1, plevel: 2, mlevel: 0},
+	{logN: 14, logQ: [2]int{6, 60}, n: 1, plevel: 3, mlevel: 0},
+	{logN: 14, logQ: [2]int{8, 60}, n: 1, plevel: 5, mlevel: 1},
+	{logN: 13, logQ: [2]int{4, 60}, n: 8, plevel: 2, mlevel: 0},
+	{logN: 14, logQ: [2]int{6, 60}, n: 8, plevel: 3, mlevel: 0},
+	{logN: 14, logQ: [2]int{8, 60}, n: 8, plevel: 5, mlevel: 1},
+	{logN: 13, logQ: [2]int{4, 60}, n: 64, plevel: 2, mlevel: 0},
+	{logN: 14, logQ: [2]int{6, 60}, n: 64, plevel: 3, mlevel: 0},
+	{logN: 14, logQ: [2]int{8, 60}, n: 64, plevel: 5, mlevel: 1},
+	{logN: 13, logQ: [2]int{4, 60}, n: 128, plevel: 2, mlevel: 0},
+	{logN: 14, logQ: [2]int{6, 60}, n: 128, plevel: 3, mlevel: 0},
+	{logN: 14, logQ: [2]int{8, 60}, n: 128, plevel: 5, mlevel: 1},
+}
+
+type vOLErings struct {
+	ringQ *ring.Ring
+	qDivP *big.Int
+	pDivM *big.Int
+}
+
+func newvOLErings(params benchParams) *vOLErings {
+
+	if params.mlevel >= params.plevel {
+		panic("mlevel must be strictly smaller than plevel")
+	}
+
+	var err error
+
+	N := 1 << params.logN
+
+	rings := new(vOLErings)
+
+	primes := ring.GenerateNTTPrimes(params.logQ[1], 2*N, params.logQ[0])
+
+	if rings.ringQ, err = ring.NewRing(N, primes); err != nil {
+		panic(err)
+	}
+
+	rings.qDivP = ring.NewInt(1)
+	for _, qi := range primes[params.plevel+1:] {
+		rings.qDivP.Mul(rings.qDivP, ring.NewUint(qi))
+	}
+
+	rings.pDivM = ring.NewInt(1)
+	for _, qi := range primes[params.mlevel+1 : params.plevel+1] {
+		rings.pDivM.Mul(rings.pDivM, ring.NewUint(qi))
+	}
+
+	return rings
+}
+
+type lowNormSampler struct {
+	baseRing *ring.Ring
+	coeffs   []*big.Int
+}
+
+func newLowNormSampler(baseRing *ring.Ring) (lns *lowNormSampler) {
+	lns = new(lowNormSampler)
+	lns.baseRing = baseRing
+	lns.coeffs = make([]*big.Int, baseRing.N)
+	return
+}
+
+func (lns *lowNormSampler) newPolyLowNorm(norm []uint64) (pol *ring.Poly) {
+	bound := ring.NewUint(0)
+	for _, qi := range norm {
+		bound.Add(bound, ring.NewUint(qi))
+	}
+
+	pol = lns.baseRing.NewPoly()
+
+	for i := range lns.coeffs {
+		lns.coeffs[i] = ring.RandInt(bound)
+	}
+
+	lns.baseRing.SetCoefficientsBigint(lns.coeffs, pol)
+
+	return
+}
+
+func main() {
```

Comment:
The example needs to be added to the `Makefile`'s `test_example` target.

Reply:
Right!